pkg/cloud/vsphere/services/govmomi/esxi/clone.go

```diff
@@ -17,12 +17,271 @@ limitations under the License.
+	ctx = context.NewMachineLoggerContext(ctx, "esxi")
+	ctx.Logger.V(6).Info("starting clone process")
+	src, err := template.FindTemplate(ctx, ctx.MachineConfig.MachineSpec.VMTemplate)
+	if err != nil {
+		return nil, err
+	}
+
+	host, err := src.HostSystem(ctx)
+	if err != nil {
+		return nil, errors.Wrap(err, "hostSystem failed")
+	}
+
+	datastore, err := ctx.Session.Finder.DatastoreOrDefault(ctx, ctx.MachineConfig.MachineSpec.Datastore)
+	if err != nil {
+		return nil, errors.Wrapf(err, "unable to get datastore for machine %q", ctx)
+	}
+
+	ctx.Logger.V(2).Info("attempting to use host resource pool")
+	pool, err := host.ResourcePool(ctx.Context)
+	if err != nil {
+		return nil, errors.Wrap(err, "host resource pool failed")
+	}
+
+	dc, err := ctx.Session.Finder.DefaultDatacenter(ctx.Context)
+	if err != nil {
+		return nil, err
+	}
+
+	vmCtx := VmContext{
+		MachineContext: ctx,
+		Datastore:      datastore,
+		Pool:           pool,
+		Datacenter:     dc,
+		Host:           host,
+		Src:            src,
+	}
+	return &vmCtx, nil
+}
+
+// GetVirtualMachineMO is a convenience method that wraps fetching the VirtualMachine
+// MO from its higher-level object.
+func (ctx *VmContext) GetVirtualMachineMO() (*mo.VirtualMachine, error) {
+	var props mo.VirtualMachine
+	if err := ctx.Src.Properties(ctx.Context, ctx.Src.Reference(), nil, &props); err != nil {
+		return nil, err
+	}
+	return &props, nil
+}
+
+// GetNics creates and returns network devices from those specifid under Network.Devices
+func GetNics(ctx *VmContext) ([]types.BaseVirtualDeviceConfigSpec, error) {
+	var devices object.VirtualDeviceList
+	for _, network := range ctx.MachineConfig.MachineSpec.Network.Devices {
+		netRef, err := ctx.Session.Finder.Network(ctx.Context, network.NetworkName)
+		if err != nil {
+			return nil, err
+		}
+
+		backing, err := netRef.EthernetCardBackingInfo(ctx.Context)
+		if err != nil {
+			return nil, err
+		}
+
+		netdev, err := object.EthernetCardTypes().CreateEthernetCard("vmxnet3", backing)
+		if err != nil {
+			return nil, err
+		}
+		devices = append(devices, netdev)
+	}
+	return devices.ConfigSpec(types.VirtualDeviceConfigSpecOperationAdd)
+}
+
+// GetSerial creates and returns a new serial port to help with entropy
+func GetSerialPort(ctx *VmContext, devices object.VirtualDeviceList) (object.VirtualDeviceList, error) {
+	ctx.Logger.V(6).Info("Adding SIO controller")
+	c := &types.VirtualSIOController{}
+	c.Key = -301
+	devices = append(devices, c)
+
+	ctx.Logger.V(6).Info("Adding Serial Port")
+
+	portSpec, err := devices.CreateSerialPort()
+	if err != nil {
+		return nil, errors.Wrapf(err, "Failed to add serial port")
+	}
+	portSpec.Key = -302
+	portSpec.ControllerKey = c.Key
+	devices = append(devices, portSpec)
+
+	return devices, nil
+}
+
+// GetVirtualDisks returns the list of existing virtual disks on the source template
+func GetVirtualDisks(ctx *VmContext) ([]types.VirtualDisk, error) {
```

Comment:
Hi @moshloop,

Please see my earlier questions about whether things need to be exported.

Also, perhaps `s/GetVirtualDisks/getDiskSpec` to be consistent with the vCenter package? Or will there be more than one disk?